docs/architecture/adr-048-consensus-fees.md

```diff
@@ -0,0 +1,169 @@
+# ADR 048: Multi Tire Gas Price System
+
+## Changelog
+
+- Dec 1, 2021: Initial Draft
+
+## Status
+
+Proposed
+
+## Abstract
+
+This ADR describes a mechanism to maintain a consensus level multi-tiered gas prices.
+
+## Context
+
+Currently, each validator configures it's own `minimal-gas-prices` in `app.yaml`. But setting a proper minimal gas price is critical to protect network from dos attack, and it's hard for all the validators to pick a sensible value, so we proposal to maintain the gas price in consensus level.
+
+Since tendermint 0.35 has supported mempool prioritization, we can take advantage of that to implement more sophisticated gas fee system.
+
+## Multi-Tier Price System
+
+We propose a multi-tier price system on consensus to provide maximum flexibility:
+
+- Tier 1: a constant gas price, which could only be modified occasionally through governance proposal.
+- Tier 2: a dynamic gas price which is adjusted according to previous block load.
+- Tier 3: a dynamic gas price which is adjusted according to previous block load at a higher speed.
```

Comment:
let's clarify how the tires will be set. I propose to make it configurable (both number and size of a tire), ideally through  gov param.

Reply:
yeah, the constant gas price is a special case of dynamic gas price, and with a single dynamic tier, it should be similar to eip-1559 already.
The main difference between this proposal and a eip-1559 design is the tx prioritization:
- in eip-1559, the user specifies a tip price to control the tx prioritization.
- while in this proposal, the user specifies the tier.
Do you think we can generalize this too?

Reply:
I think it's already general enough. I propose to clarify how the amount of tires will be set and if we can configure it once the chain is running (I propose yest - through the gov param).

Reply:
done, add a parameter schema, the `tier` field should be simply the index to the tier list.